docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,129 @@
+
+## Abstract
+
+A proposal for improving Kubelet config by versioning it and making it manageable via the API server.
+
+## Motivation
+
+The Kubelet is currently configured via command-line flags. This is painful for a number of reasons:
+- It makes it difficult to change the way Kubelets are configured in a running cluster, because it is often tedious to change the Kubelet startup configuration.
+- It makes it difficult to manage different Kubelet configurations for different nodes, e.g. if you want to canary a new config.
+- The current lack of versioned Kubelet configuration means that any time we change Kubelet flags, we risk breaking someone's setup.
+- Having the ability to pass config in the same format on-disk as via the API server will make it easier to bootstrap master nodes before the API server is available.
+
+
+## Goals of this design:
+
+- Add a well-defined, versioned configmap object for Kubelet configuration.
+- Add/Remove/Update global and per-node Kubelet config on the fly via the API server.
+- Provide a reliable way to get the configuration of a given Kubelet along with its source (remote or on disk or flags).
+- We want the Kubelet to work with the built-in defaults on all distros we test against.
+- Work towards deprecating flags in favor of on-disk config and config from the API server. Master nodes will be bootstrapped from on-disk config.
+- Make it possible to opt-out of remote configuration e.g. to protect the master from bad config updates.
+
+## Design
+
+### Definitions
+
+- `kubelet-default`: The built-in default config that every Kubelet starts up with.
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. Please note that `v1alpha1` will not be regrouped, but eventually we will start working on this.
+
+@timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
```

Comment:
First, the current componentconfig is just a stepping stone. I recommend we don't exit alpha until we at least have created types for all components, group related fields, and decide for each field whether we really want to keep it, in its current form. Whether the reorganized config types should be v1alpha2 or v1beta1 depends on our level of confidence in our ability/desire to maintain backward compatibility. 

Second, the model of evolution that I think would work best is to use an incubation group, where new APIs would start until they reach the level of maturity of the group in which they should eventually land. extensions is effectively our incubation group for actual REST APIs. We could create a componentincubation  group or somesuch for componentconfig schemas.